### ISSUE TITLE
Page linking Race Condition

### ISSUE THREAD

**Issue by doctormo** [jpakkane/capypdf#58]
Because we can't paint pages *after* adding them, we don't have ids to use in Link annotations.

We need a way to reserve ids for pages such that they *will have* this id when serialised, but haven't been serialised yet.

Without this it will be impossible to make links between pages.

(Note: Cairo also have this bug, as it's not possible to link to future pages in cairo either)

**Comment by doctormo** [comment]
Whatever solution we come up with should also work for `12.3.2.3 Named Destinations` which might also be created before or after they are used in a Link annotation.

**Comment by jpakkane** [comment]
For pages at least you should be able to give it any physical page number. At generation time it would be validated to point to an existing page or error out.

**Comment by doctormo** [comment]
Can you explain how that would work? We'd print `/Dest [11 0 R ]` which points to object `11 0` which doesn't exist when the Annotation is created and added to the document (and then used in links in the document). So how do we know it's `11 0 R` ? It won't have the data yet.

**Comment by jpakkane** [comment]
I have not looked at the code in detail but basically the same way as every other case where this happens. You put a `DelayedThingamabob` in the object array and then in `PdfWriter` convert that to the final PDF dict.

**Comment by doctormo** [comment]
The page already appears to be constructing a `DelayedPage` object. Although it's lifetime is unclear as it seems to seal all the useful parts such as the command_stream immediately. So what I need is something like a `PromiseToMakeAPage` which can give me a PageId and I can then use that in Annotations and have it error out if I try and write the pdf before I've constructed all the promised pages.

But maybe I've got this the wrong way around and I still don't understand what you're thinking of having for this use case.
